rfd/0150-ssh-connection-resumption.md

```diff
@@ -0,0 +1,129 @@
+
+## Description of the protocol
+
+In the following description, all numbers are sent as variable length unsigned integers, as per [Protobuf specs](https://protobuf.dev/programming-guides/encoding/#varints), or as implemented by the Go standard library in [`encoding/binary`](https://pkg.go.dev/encoding/binary).
+
+### Version exchange
+
+The client blindly sends the bytestring `SSH-2.0-`. As this is 8 bytes, it's enough to trigger the Teleport protocol multiplexer (which unfortunately blocks until the client sends 8 bytes), which will direct the connection to the resumption protocol handler, which will send the bytestring `teleport-resume-v0\r\n`. If the client receives such a string from the server, it will then send `\x00teleport-resume-v0` (the NUL is explicitly disallowed by the SSH version handshake, see [RFC 4253 section 4.2](https://datatracker.ietf.org/doc/html/rfc4253#section-4.2)), and both ends are in agreement about the use of the "resumption protocol v0" from that point of the connection.
+
+If the first bytes sent by the server don't match `teleport-resume-v0\r\n` then the client must assume that the server is a regular SSH server, and thus will abort the resumption protocol version exchange, and begin a normal SSH client connection; as `SSH-2.0-` was already sent in the connection, eight bytes must be skipped, assuming that the SSH client will send the same. Such an assumption is valid, as the client side of an SSH connection must send `SSH-2.0-<softwareversion>` at the very beginning of the connection.
+
+If the first bytes sent by the client don't match `SSH-2.0-\x00teleport-resume-v0` then the server must assume that the client is a regular SSH client, and thus will abort the resumption protocol version exchange, and begin a normal SSH server connection. No special handling of the connection is required here, as the server side of the exchange is allowed to send CRLF-terminated lines of UTF-8 text that don't begin with `SSH-` and that don't contribute to the SSH connection state.
+
+### Handshake (new connection)
+
+The clients sends `\x00` (likely together with the previous data), and the server generates a _resumption token_ out of 16 random bytes, but with the highest bit of the first byte set. The server sends this resumption token, then its length-prefixed host ID (which is almost always a UUID in string form, but can also be an EC2 instance ID of varying length). The host ID of the server should be used in later reconnection attempts to ensure that the correct server is receiving the new connection, since the first connection might've been to a hostname or public address and thus might eventually resolve to a different server.
```

Comment:
We can do that, that should save us either half or one roundtrip. If we went back to my original plan of using the `comments` of the server side version exchange rather than a line before that we could save even slightly more, if the client can make use of that data immediately.

Reply:
It seems worth the effort to reduce as many RTT's as possible if we are trying to minimize how noticeable the resumption is.

Reply:
Added in c5f7f912a296b3914bfe6773c5eea8702938769c